zenoh/src/lib.rs

```diff
@@ -364,26 +355,58 @@ pub mod config {
     pub use zenoh_config::*;
 }
 
-#[doc(hidden)]
-#[cfg(all(feature = "unstable", feature = "plugins"))]
-pub mod plugins {
-    pub use crate::api::plugins::{
-        PluginsManager, Response, RunningPlugin, RunningPluginTrait, ZenohPlugin, PLUGIN_PREFIX,
-    };
-}
+#[cfg(all(feature = "internal", not(feature = "unstable")))]
+compile_error!(
+    "All internal functionality is unstable. The `unstable` feature must be enabled to use `internal`."
+);
+#[cfg(all(
+    feature = "plugins",
+    not(all(feature = "unstable", feature = "internal"))
+))]
+compile_error!(
+    "The plugins support is internal and unstable. The `unstable` and `internal` features must be enabled to use `plugins`."
+);
 
-#[doc(hidden)]
+#[zenoh_macros::internal]
 pub mod internal {
     pub use zenoh_core::{zasync_executor_init, zerror, zlock, ztimeout, ResolveFuture};
     pub use zenoh_result::bail;
     pub use zenoh_sync::Condition;
     pub use zenoh_task::{TaskController, TerminatableTask};
     pub use zenoh_util::{zenoh_home, LibLoader, Timed, TimedEvent, Timer, ZENOH_HOME_ENV_VAR};
 
-    pub use crate::api::encoding::EncodingInternals;
+    /// A collection of useful buffers used by zenoh internally and exposed to the user to facilitate
+    /// reading and writing data.
+    pub mod buffers {
+        pub use zenoh_buffers::{
+            buffer::SplitBuffer,
+            reader::{HasReader, Reader},
+            ZBuf, ZBufReader, ZSlice, ZSliceBuffer,
+        };
+    }
+    /// Initialize a Session with an existing Runtime.
+    /// This operation is used by the plugins to share the same Runtime as the router.
+    #[zenoh_macros::internal]
+    pub mod runtime {
+        pub use zenoh_runtime::ZRuntime;
+
+        pub use crate::net::runtime::{AdminSpace, Runtime, RuntimeBuilder};
+    }
+    /// Plugins support
+    #[cfg(feature = "plugins")]
+    pub mod plugins {
+        pub use crate::api::plugins::{
+            PluginsManager, Response, RunningPlugin, RunningPluginTrait, ZenohPlugin, PLUGIN_PREFIX,
+        };
+    }
 }
 
-#[cfg(all(feature = "unstable", feature = "shared-memory"))]
+#[cfg(all(feature = "shared-memory", not(feature = "unstable")))]
```

Comment:
Do you want to require user putting both `shared_memory` and `unstable` features in their `Cargo.toml`, or could `unstable` be a dependency of `shared_memory`? I'm ok with the first solution.

Reply:
Same as above, the "unstable" feature is a red flag for user, he should explicitly agree, that he is stepping on shaky ground

Reply:
If it is such a red flag, you may be interested in [`tokio` approach](https://docs.rs/tokio/latest/tokio/#unstable-features).

Reply:
As far as I see our approach is exactly the same: "some feature flags and some api parts are available only if `unstable` flag is set"

Reply:
Our approach uses compilation feature, while tokio is using compiler *cfg* flag. This is radically different, as explained in the [post](https://internals.rust-lang.org/t/feature-request-unstable-opt-in-non-transitive-crate-features/16193#why-not-a-crate-feature-2) linked in tokio documentation quoted above.
I honestly think we should follow the same convention, but that's for an other PR. I will open an issue to discuss it.